### ISSUE TITLE
Dictionary strong key names

### ISSUE THREAD

**Issue by MohammadHamdyGhanem** [dotnet/vblang#234]
I suggest to add this form to the dictionary initializer:
```
Dim Cities As New Dictionary(Of String, Integer)
        From {NewYork = 10, London = 5, Cairo = 12}
```
Note that:
1- I Suggest also to add an implicit line continuation arround the From keyword.
2- No quotations arround key names, not only for simplicity, but also to force us to choose valid names to be used as members (see 3).
3- The complier can use this syntax to define strong key names seen only in the dictionary scope so they don't affect other variable names. These keys can be used as follows:
```
Msgbox(Cities.Cairo) 
Cities.London = 3
```
Or by the traditional ways:
```
Msgbox(Cities!Cairo) 
Cities(London) = 3
```
This will prevent key names mistyping bugs, and give us intillisense autocompletion.

**Comment by reduckted** [comment]
If you know the names of the keys (and this would only work with string keys), wouldn't an anonymous type or tuple be a better choice than a dictionary?

**Comment by MohammadHamdyGhanem** [comment]
`If you know the names of the keys (and this would only work with string keys), wouldn't an anonymous type or tuple be a better choice than a dictionary?`

Anonymous types and tuples can't have new members at run time, and can't be enumerated easily. 
When you use dictionaries, some keys are known and hard coded, but other keys can be added in run-time.

**Comment by pricerc** [comment]
> Anonymous types and tuples can't have new members at run time, and can't be enumerated easily.
When you use dictionaries, some keys are known and hard coded, but other keys can be added in run-time.

how would
>`Msgbox(Cities!Cairo)`
>`Cities(London) = 3`

work with data added at run-time?

> If you know the names of the keys (and this would only work with string keys), wouldn't an anonymous type or tuple be a better choice than a dictionary?

to me, it looks more like an enum

**Comment by MohammadHamdyGhanem** [comment]
> work with data added at run-time?

We care about writing code in design time. You can't know the key that will be added in run-time, so you don't write them in code. If you expect a key to be added, then you can add it to the initializer with a default value.

**Comment by reduckted** [comment]
> When you use dictionaries, some keys are known and hard coded, but other keys can be added in run-time.

Do you have a practical example of this?

**Comment by AnthonyDGreen** [comment]
Best I think I could do for you is:

``` VB.NET
Dim cities = New Dictionary(Of String, Integer)
    With { !NewYork = 10, !London = 5, !Cairo = 12 }
```

Allowing you to use the existing dictionary access operator `!` in an object initializer in the way you can use member access `.` today.

**Comment by AnthonyDGreen** [comment]
I have another idea to run by you to enable to strong typing.

**Comment by MohammadHamdyGhanem** [comment]
The problem exists even with the  !  syntax. If you mistype a key name, it is consederd a new key and added to the dictionary. This leads to unwanted, hard to discover results!
Maybe it wlll be better if the dictionary raises an exception if the key doesn't exist. Then, the need to hard typing will be minimal.

**Comment by MohammadHamdyGhanem** [comment]
@AnthonyDGreen 
Regarding dictionaries:
I think of Tuples and the upcoming Records as a specialized collection of elements of different types. I think they should implement the IDictionary<string, T> interface and have a ToDictionary<string, T> method. T can be object if elements have different types,or can be a specific type if all elements have the same type. This wsy we can use the Typle or the Record as a stronge key dictionary,  having the ability to enumeration and can convert it to a dictionary any time to pass it to legacy methods or allow adding keys at runtime. Even better if there is a way to generate a tuple from a dictionary.

**Comment by pricerc** [comment]
@MohammadHamdyGhanem 
I think I phrased my question poorly.

In your initial proposal, you describe a compile-time, strongly typed, 'intellisensible' dictionary, and make no mention of adding things at run-time. I can see some value in this as a mechanism for 'structured constants' where you're doing calculations using different parameters based on some key value.

But then, in your response, to @reduckted, you observed:

> other keys can be added in run-time.

I don't see this as compatible with your initial description, which is why I asked how your sample code would work for dynamically added dictionary members. I think @reduckted phrased it better by asking for a practical example.

Also, if keys can be added at run-time, then presumably, they can also be removed. So what happens if an entry is deleted from the dictionary:
`
Dim Cities As New Dictionary(Of String, Integer)
        From {NewYork = 10, London = 5, Cairo = 12}

Cities.Delete(Cities.Cairo)
`
?

**Comment by MohammadHamdyGhanem** [comment]
> So what happens if an entry is deleted from the dictionary:

Current behavior makes the dictionary re-add the removed key if you try to access it again:

> If you mistype a key name, it is consederd a new key and added to the dictionary. This leads to unwanted, hard to discover results!

**Comment by MohammadHamdyGhanem** [comment]
I proposed a new solution #235

**Comment by bandleader** [comment]
@MohammadHamdyGhanem Please click "Close issue" on the bottom of the page.
Also, I replied to #235.

**Comment by bandleader** [comment]
@MohammadHamdyGhanem Kindly click "Close issue" on the bottom of the page, to reduce clutter here.
Also, I replied to #235.

**Comment by tverweij** [comment]
@MohammadHamdyGhanem 
I think this is what you want:
```
        Const NewYork As String = "NewYork"
        Const London As String = "London"
        Const Cairo As String = "Cairo"

        Dim Cities As New Dictionary(Of String, Integer) From {{NewYork, 10}, {London, 5}, {Cairo, 12}}
        MsgBox(Cityies(NewYork))
```
Or do I misunderstand the problem?

**Comment by AdamSpeight2008** [comment]
@MohammadHamdyGhanem  Have you also looked at `ExpandoObject`?

**Comment by MohammadHamdyGhanem** [comment]
@AdamSpeight2008 
Of course. It has the same problems , that makes me search for a new solution to make it strong types. This is why I proposed Typed Dictionary. A Typed ExpandoObject can be done the same way, because it is a dictionary too.
